src/Icicle/Avalanche/Statement/Simp.hs

```diff
@@ -318,10 +316,109 @@ substXinS
  :: (Hashable n, Eq n)
  => a
  -> Name n
+ -> Exp a n p
+ -> Statement a n p
+ -> Fresh n (Statement a n p)
+substXinS a_fresh name payload statements
```

Comment:
you may get some benefit from making this take a substitution map rather than a single substitution, I'm not sure